api/src/services/folders.ts

```diff
@@ -1,8 +1,76 @@
-import type { AbstractServiceOptions } from '@directus/types';
+import type { AbstractServiceOptions, Folder } from '@directus/types';
+import { validateAccess } from '../permissions/modules/validate-access/validate-access.js';
+import { NameDeduper } from './assets/name-deduper.js';
 import { ItemsService } from './items.js';
 
-export class FoldersService extends ItemsService {
+export class FoldersService extends ItemsService<Folder> {
 	constructor(options: AbstractServiceOptions) {
 		super('directus_folders', options);
 	}
+
+	async tree(root: string) {
+		if (this.accountability && this.accountability.admin !== true) {
+			await validateAccess(
+				{
+					collection: 'directus_folders',
+					accountability: this.accountability,
+					action: 'read',
+					primaryKeys: [root],
+				},
+				{
+					knex: this.knex,
+					schema: this.schema,
+				},
+			);
+		}
+
+		const folders = await this.readByQuery({});
+
+		// build folder and child lookup
+		const folderLookup = new Map<string, Folder>();
+		const childFolderLookup = new Map<string, string[]>();
+
+		for (const folder of folders) {
+			if (!folder['id']) continue;
```

Comment:
This can happen as it's a primary key ?

Reply:
Although it’s unlikely, policy rules do allow denying access to `id`. Since we’re using `readByQuery`, it will return `null` or `none` if access is denied.